### PR TITLE
Issue 14

### LINKED ISSUE #14: Write tests for hocon function
Need unit tests for functions in hocon parser part. Need using special structures writed for unit test in any sources for configuration structures

### PR DESCRIPTION
**Pull request by kubitre**
resolve #14 
+ add tests for pipeline function
+ add tests for tools
+ add tests for lib

### CODE REVIEW

File: pipeline/pipeline.go [hunk 1]
```diff
@@ -261,37 +276,31 @@ func (pipeline *Pipeline) configuringValues(context *structContext) error {
...
+		// add check for setuping valueGet in valueIndirect
+		if valueIndirect.CanSet() {
+			fmt.Println("Loglevel: Debug Message: [Pipeline]: setupe value in struct")
+			valueIndirect.Set(value.Value)
+			return nil
+		} else {
+			return errors.New("can not set " + value.Value.Kind().String() + " into struct field.")
+		}
+	} else {
+		return errors.New("Loglevel: Debug Message:  value get not implementedable value: ")
```
Comment (artemkaxboy) [inline pipeline/pipeline.go:302]: I believe there must be error level not debug

File: tags/tags.go [hunk 2]
```diff
@@ -11,7 +11,7 @@ const (
 	TagConfigServer   = prefixLibTag + "server"
 	TagHashiCorpVault = prefixLibTag + "vault"
 
-	AmountTags = 7
+	AmountTags = 8
```
Comment (artemkaxboy) [inline tags/tags.go:14]: You have not added anything here, why do you increase the number?

**Review by artemkaxboy — CHANGES_REQUESTED** [review]

File: lib_test.go [hunk 3]
```diff
@@ -23,16 +23,24 @@ type (
...
+	// 	NestedStruct2 struct {
+	// 		Field1 string `cf_hocon:"test1"`
+	// 	} `cf_hocon:"node=planB"`
+	// }
+
+	// MyStruct4 struct {
+	// 	NestedStruct4 struct {
+	// 		Field1 string
+	// 	} `cf_hocon:"node=planC.tururu.tratatat.planZ"`
+	// }
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:36]: yes, this need repair in latest issues

File: lib_test.go [hunk 4]
```diff
@@ -66,40 +74,94 @@ func Test_parseHocon(t *testing.T) {
...
+// 	myStruct, err := ConfigureSmart(&MyStruct4{}, "./test_configs/testmap.hocon")
+// 	if err != nil {
+// 		fmt.Println("error while configuring: ", err)
+// 	}
+// 	assert.Equal(t, &MyStruct4{
+// 		NestedStruct4: struct{ Field1 string }{
+// 			Field1: "testValueByTest",
+// 		},
+// 	}, myStruct.(*MyStruct4))
+// }
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:116]: no

File: lib_test.go [hunk 5]
```diff
@@ -66,40 +74,94 @@ func Test_parseHocon(t *testing.T) {
...
+		Field2: "test",
+		Field3: true,
+		Field4: 12.2,
+		Field5: []bool{true, false, true},
+	}, myStruct.(*EnvStruct))
+	os.Remove("myField1")
+	os.Remove("myField2")
+	os.Remove("myField3")
+	os.Remove("myField4")
+	os.Remove("myField5")
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:141]: why? https://github.com/goreflect/gostructor/pull/22#discussion_r403694190

File: lib_test.go [hunk 6]
```diff
@@ -66,40 +74,94 @@ func Test_parseHocon(t *testing.T) {
...
+	os.Setenv("myField1", "12")
+	os.Setenv("myField2", "test")
+	os.Setenv("myField3", "true")
+	os.Setenv("myField4", "12.2")
+	os.Setenv("myField5", "true,false,true")
+
+	myStruct, err := ConfigureSmart(&EnvStruct{}, "")
 	if err != nil {
-		fmt.Println("error while configuring: ", err)
+		t.Error(err)
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:127]: if will be error this method closing automatic and return code of error. Check please

File: lib_test.go [hunk 7]
```diff
@@ -66,40 +74,94 @@ func Test_parseHocon(t *testing.T) {
...
+func Test_configureEasy(t *testing.T) {
+	os.Setenv("myField1", "12")
+	os.Setenv("myField2", "test")
+	os.Setenv("myField3", "true")
+	os.Setenv("myField4", "12.2")
+	os.Setenv("myField5", "true,false,true")
+	myStruct, err := ConfigureEasy(&EnvStruct{}, "")
 	if err != nil {
-		fmt.Println("error while configuring: ", err)
+		t.Error(err)
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:152]: https://github.com/goreflect/gostructor/pull/24#discussion_r403694962

File: lib_test.go [hunk 8]
```diff
@@ -66,40 +74,94 @@ func Test_parseHocon(t *testing.T) {
...
+		Field2: "test",
+		Field3: true,
+		Field4: 12.2,
+		Field5: []bool{true, false, true},
+	}, myStruct.(*EnvStruct))
+	os.Remove("myField1")
+	os.Remove("myField2")
+	os.Remove("myField3")
+	os.Remove("myField4")
+	os.Remove("myField5")
```
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:166]: https://github.com/goreflect/gostructor/pull/24#discussion_r403694752

File: pipeline/environment_function.go [hunk 9]
```diff
@@ -3,21 +3,58 @@ package pipeline
...
-	fmt.Println("Level: Debug. Message: environment values sources start")
+	fmt.Println("Level: Debug. Environment values sources start")
+	valueIndirect := reflect.Indirect(context.Value)
+	valueTag := context.StructField.Tag.Get(tags.TagEnvironment)
+
+	if valueTag != "" {
+		value := os.Getenv(valueTag)
+		return converters.ConvertBetweenPrimitiveTypes(reflect.ValueOf(value), valueIndirect)
+	}
+	return infra.NewGoStructorNoValue(context.Value, errors.New("getBaseType can not getting field by empty tag value of tag: "+tags.TagEnvironment))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply pipeline/environment_function.go:53]: +=

File: pipeline/environment_function_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+			}
+		})
+	}
+}
+
+func TestEnvironmentConfig_GetComplexType(t *testing.T) {
+	myStruct := struct {
+		fieldhard []float32 `cf_env:"mySlice"`
+	}{}
+	os.Setenv("mySlice", "1.12,1.24,1.67")
```
[the remark this answers is not in the document]
Comment (kubitre) [reply pipeline/environment_function_test.go:98]: this fixed in https://github.com/goreflect/gostructor/pull/22#discussion_r403694260

File: pipeline/pipeline.go (hunk 1 above)
Comment (kubitre) [reply pipeline/pipeline.go:302]: for all logs need refactore by resolve this issue: https://github.com/goreflect/gostructor/issues/12

File: tags/tags.go (hunk 2 above)
Comment (kubitre) [reply tags/tags.go:14]: Max index for last tag = 7. If i will create array for len = 7 program will exit with panic. This need for logic part programm. I can not be changed this

File: lib_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (artemkaxboy) [reply lib_test.go:133]: No it will not
image

[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:133]: ++

File: lib_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:153]: ++

File: lib_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kubitre) [reply lib_test.go:167]: ++

**Pull request merged**